### DISCUSSION TITLE
How to extract few files from output channel which names match a string?

### DISCUSSION THREAD

**Discussion by zillurbmb51** [nextflow-io/nextflow#2011] [Q&A]
Hi, I have a process that generates multiple output. I need few files in one of the subsequent processes. 
```
output:
tuple val(chrom), path("chr${chrom}*.{bed,bim,fam}") into qc1_chr
```
from qc1_chr I need some files which names match a string (say, qc1_chr has 200 files, and only 22 file names have a string 'step10').
I tried with 'filter' operator but failed. 
`qc1_chr.filter(~/*step10*/) into qc2`  But it does not work.
How could I extract those files into a new channel? Thanks in advance. Best, Zillur

**Comment by phue** [comment]
Hi,
your output channel is a tuple of 2, so you'll probably want to use `.filter` with a closure:
`qc2 = qc1_chr.filter { it[1].name.contains('step10') }`

If you want to keep all files, you could split them into separate channels using the `.branch` operator:

```
qc1_chr
  .branch{ chrom, filename ->
      match: filename.name.contains('step10')
          return [chrom, filename]
      nomatch: true
          return [chrom, filename]
  }
  .set {qc2}

qc2.match.view()
```

**Reply by zillurbmb51** [reply]
Thanks @phue . But none of these two works! Both codes generate no output. But qc1_chr output channel has all the files `[chr1-22,[respective files ]]`

**Reply by phue** [reply]
I did not realize that your channel contains a list of filepaths.

Maybe you can just emit both channels from your process?
```
output:
tuple val(chrom), path("*step10*.{bed,bim,fam}") into qc2
tuple val(chrom), path("chr${chrom}*.{bed,bim,fam}") into qc1
```

**Comment by zillurbmb51** [comment]
Thanks @phue . The following code works for me.
`qc1_chr.map{chrom,bfiles -> tuple(chrom,bfiles.findAll{it.baseName.endsWith('step10')})}`

**Comment by slsevilla** [comment]
Needing some extra guidance on filtering an entire tuple based on one value. I used this discussion to get part-way. I have a channel of tuples (reads_ch):
```
[[id:S1, single_end:false], [/home/ubuntu/S1.R1.fastq.gz, /home/ubuntu/S1.R2.fastq.gz], /home/ubuntu/S1.txt, [PASSED: File S1_R1 is not corrupt., PASSED: File S1_R2 is not corrupt.]]
[[id:S2, single_end:false], [/home/ubuntu/S2.fastq.gz, /home/ubuntu/S2.R2.fastq.gz], /home/ubuntu/S2.txt, [PASSED: File S2_R1 is not corrupt., PASSED: File S2_R2 is not corrupt.]]
[[id:S3, single_end:false], [/home/ubuntu/S3.R1.fastq.gz, /home/ubuntu/S3.R2.fastq.gz], /home/ubuntu/S3.txt, [FAILED, FAILED]]
```

I am trying to filter this for any sample that did not 'fail'. I've gotten as far as clearing it[3] but haven't figured out how to remove that sample. while keeping the structure of the remaining samples (S1,S2)

Goal: 
```
[[id:S1, single_end:false], [/home/ubuntu/S1.R1.fastq.gz, /home/ubuntu/S1.R2.fastq.gz], /home/ubuntu/S1.txt, [PASSED: File S1_R1 is not corrupt., PASSED: File S1_R2 is not corrupt.]]
[[id:S2, single_end:false], [/home/ubuntu/S2.fastq.gz, /home/ubuntu/S2.R2.fastq.gz], /home/ubuntu/S2.txt, [PASSED: File S2_R1 is not corrupt., PASSED: File S2_R2 is not corrupt.]]
```

Attempt: 
```
passing = read_ch.map { meta, reads, outcome_file, outcome_status ->
tuple(meta, reads, outcome_file, outcome_status.findAll { !it.contains('FAILED') })
}
passing.view()

[[id:S1, single_end:false], [/home/ubuntu/S1.R1.fastq.gz, /home/ubuntu/S1.R2.fastq.gz], /home/ubuntu/S1.txt, [PASSED: File S1_R1 is not corrupt., PASSED: File S1_R2 is not corrupt.]]
[[id:S2, single_end:false], [/home/ubuntu/S2.fastq.gz, /home/ubuntu/S2.R2.fastq.gz], /home/ubuntu/S2.txt, [PASSED: File S2_R1 is not corrupt., PASSED: File S2_R2 is not corrupt.]]
[[id:S3, single_end:false], [/home/ubuntu/S3.R1.fastq.gz, /home/ubuntu/S3.R2.fastq.gz], /home/ubuntu/S3.txt, []]
```
